# Assignment 1: Part B (55/100 marks)

We're moving to New York City!  We need to find a place to live, so let's get acquainted with our new home.  

## NYC OpenData - 311 Service Requests for 2010
<em>"Open Data is an opportunity to engage New Yorkers in the information that is produced and used by City government. We believe that every New Yorker can benefit from Open Data, and Open Data can benefit from every New Yorker."</em>(https://opendata.cityofnewyork.us/overview/)

Data: https://data.cityofnewyork.us/Social-Services/311-Service-Requests-for-2009/3rfa-3xsf


In [5]:
import pandas as pd
import numpy as np

class display(object):
    """Display HTML representation of multiple objects"""
    template = """<div style="float: left; padding: 10px;">
    <p style='font-family:"Courier New", Courier, monospace'>{0}</p>{1}
    </div>"""
    def __init__(self, *args):
        self.args = args
        
    def _repr_html_(self):
        return '\n'.join(self.template.format(a, eval(a)._repr_html_())
                         for a in self.args)
    
    def __repr__(self):
        return '\n\n'.join(a + '\n' + repr(eval(a))
                           for a in self.args)

## IMPORTANT: 
The `complaints_df` DataFrame created below is a global variable. You can use the `complaints_df` DataFrame to access the relevant info in each of your question functions.  But do not modify the DataFrame.

The `311_Service_Requests_2010.csv` data file has over a million rows, so the file may take a few minutes to load!

In [6]:
complaints_df = pd.read_csv('data/311_Service_Requests_2010.csv', 
                            dtype = 'str',
                            parse_dates = ['Created Date'])

### Complaints data 
The next few blocks, provide an overview of the 311 Service Requests for 2010.

- `complaints_df.describe()`: provides summary statistics
- `complaints_df.head()`: displays the first 5 rows of the DataFrame.  Note the various column labels.  Review the `311_SR_Data_Dictionary_2018.xlsx` file, provided in your assignment's data directory. In the 311_SR_Data_Dictionary_2018, you’ll find definitions of terms and values describing each column's data.
- `complaints_df.shape`: The number of rows and columns in your dataset.  The `311_Service_Requests_2010.csv` includes all of the rows  where the Borough was specified. Columns which were not required for our analysis, were removed to reduce the data file's size.

In [7]:
complaints_df.describe()

Created Date       Complaint Type        Descriptor  \
count               1048575              1048575           1038167   
unique               670112                  180               899   
top     2010-11-04 11:39:00  Noise - Residential  Loud Music/Party   
freq                    281               104314             95099   
first   2010-01-01 00:00:00                  NaN               NaN   
last    2010-11-23 13:26:00                  NaN               NaN   

       Location Type Incident Zip Street Name Cross Street 1 Cross Street 2  \
count         713734      1044330      706441         745705         745348   
unique           104          203       10035          14545          14796   
top           Street        11207    BROADWAY           BEND           BEND   
freq          203358        13821        7347          12549          13038   
first            NaN          NaN         NaN            NaN            NaN   
last             NaN          NaN         NaN            NaN            NaN   

            City   Status Community Board   Borough  
count    1044549  1048575         1048575   1048575  
unique        89        8              71         5  
top     BROOKLYN   Closed       12 QUEENS  BROOKLYN  
freq      306981   965600           35486    308127  
first        NaN      NaN             NaN       NaN  
last         NaN      NaN             NaN       NaN

In [8]:
complaints_df.head()

Created Date        Complaint Type  \
0 2010-01-01 00:00:00               HEATING   
1 2010-01-01 00:00:00  GENERAL CONSTRUCTION   
2 2010-01-01 00:00:00  GENERAL CONSTRUCTION   
3 2010-01-01 00:03:00   Noise - Residential   
4 2010-01-01 00:04:00          Building/Use   

                                          Descriptor  \
0                                               HEAT   
1                                              DOORS   
2                                               MOLD   
3                                    Loud Television   
4  SRO - Illegal Work/No Permit/Change In Occupan...   

                Location Type Incident Zip       Street Name   Cross Street 1  \
0        RESIDENTIAL BUILDING        10468  RESERVOIR AVENUE  WEST 195 STREET   
1        RESIDENTIAL BUILDING        10468  RESERVOIR AVENUE  WEST 195 STREET   
2        RESIDENTIAL BUILDING        10468  RESERVOIR AVENUE  WEST 195 STREET   
3  Residential Building/House        11230    EAST 19 STREET         AVENUE O   
4                         NaN        10466  EAST  224 STREET             BEND   

       Cross Street 2      City  Status Community Board   Borough  
0      GOULDEN AVENUE     BRONX    Open        08 BRONX     BRONX  
1      GOULDEN AVENUE     BRONX    Open        08 BRONX     BRONX  
2      GOULDEN AVENUE     BRONX    Open        08 BRONX     BRONX  
3            AVENUE P  BROOKLYN  Closed     14 BROOKLYN  BROOKLYN  
4  SCHIEFFELIN AVENUE     BRONX  Closed        12 BRONX     BRONX

In [9]:
complaints_df.shape

(1048575, 12)

### B0 (example):  _Which Boroughs are included in our 2010 complaints database?_   

Function `b0()` should return a **Series object** with the Borough's listed in alphabetical order.

In [10]:
def b0():
    """ Returns a Series object with an alphabetical list
        of the NYC Boroughs    
    """ 
    boroughs = complaints_df['Borough'].sort_values().unique()

    return boroughs

In [11]:
b0()

array(['BRONX', 'BROOKLYN', 'MANHATTAN', 'QUEENS', 'STATEN ISLAND'],
      dtype=object)

### B1: _What are the top ten most common complaint types?_ (10 marks)

Create a new `DataFrame`from a dictionary of two series objects.  Your `DataFrame` should have columns for 'Count', and 'Percentage' with the `Complaint Type` as each row's index.  It should also contain only the top ten complaint types, ie it should only have 10 rows.
   
Count values should be integers. Percentage's should be floats, rounded to two decimal places. See example below which shows the first row.  


| &nbsp; | Count  |	Percentage |  
|------- |------- |------------|  
|Noise - Residential | 	104314 | 9.95 |  


In [12]:
def b1():
    """Returns the top ten complaints number of times filed and their percentage"""
    column1 = complaints_df['Complaint Type'].value_counts()
    column2 = 100 * column1/column1.sum()
    data = pd.DataFrame({'count':column1,'percentage':column2})[:10]
    return data

In [13]:
b1()

count  percentage
Noise - Residential            104314    9.948168
Water System                    61540    5.868917
Street Light Condition          58899    5.617052
Blocked Driveway                43986    4.194836
Street Condition                42468    4.050068
Sewer                           37989    3.622917
Damaged Tree                    37432    3.569797
Dirty Conditions                32938    3.141215
General Construction/Plumbing   29417    2.805426
Building/Use                    27759    2.647307

### B2: _What percentage of the 311 complaints are Noise related?_ (10 marks)
  
Noise related complaints are assumed to have `Noise` included in the *Complaint Type* field's text string.  Function `b2()` should return a `string` with the calculated percentage rounded to the nearest integer, such as '37%'.

*Hint: consider using Pandas string methods to find all noise related complaints. p. 180-181 PyDSHB.*

In [14]:
def b2():
    """Returns the percentage of complaint types where 'noise' is used"""
    
    noise = complaints_df['Complaint Type'].str.contains('[Nn]oise').sum()
    total = complaints_df['Complaint Type'].count()
    p = int(100 * noise / total)
    return str(p)+'%'
   

In [15]:
b2()

'17%'

### B3: _What are the ten most common Noise complaints?_ (10 marks)

Create a `DataFrame` similar to question b1()'s. Your `DataFrame` should have columns for 'Count', and 'Percentage' but the row index will use the **_Descriptor_** field, not the *Complaint Type*, as each row's label.  
  
Percentage's should be rounded to two decimal places. And the Count's should be integers. 
  
Add a final row for a category of **_Other_**, which represents all *other* noise related complaints.  See table below which provides an example for the first and last rows.

| &nbsp; | Count  |	Percentage |  
|------- |------- |------------|  
|Loud Music/Party |	95099 |	51.60  
|... | 	... | ... |  
|Other | 	10651 | 5.77 |  


In [266]:
def b3():
    """Returns most common complaints"""
    
    total = complaints_df['Descriptor'].count()
    noise = complaints_df['Descriptor'][complaints_df['Complaint Type'].str.contains('[Nn]oise')].count()
    column1 = complaints_df['Descriptor'][complaints_df['Complaint Type'].str.contains('[Nn]oise')].value_counts()
    column1 = round(column1,0)
    other = total - noise
    other_percentage = 100 * noise /total
    column2 = 100 * column1/column1.sum()
    data = pd.DataFrame({'count':column1,'percentage':column2})[:10]
    row = pd.Series({'count':other,'percentage':other_percentage},name='Other')
    data = data.append(row)
    return data
    

In [267]:
b3()

count  percentage
Loud Music/Party                                     95099.0   51.600000
Banging/Pounding                                     29140.0   15.810000
Loud Talking                                         17079.0    9.270000
Car/Truck Music                                       8168.0    4.430000
Noise: Construction Before/After Hours (NM1)          5768.0    3.130000
Noise, Barking Dog (NR5)                              5695.0    3.090000
Engine Idling                                         3535.0    1.920000
Noise: Construction Equipment (NC1)                   3477.0    1.890000
Noise: Jack Hammering (NC2)                           3315.0    1.800000
Noise: Air Condition/Ventilation Equip, Commerc...    2382.0    1.290000
Other                                               853858.0   17.753309

### Community Boards (Districts)  of New York City
The next few questions provide an overview of the New York City community districts.  New York City is divided into five boroughs: 'BRONX', 'BROOKLYN', 'MANHATTAN', 'QUEENS',  and 'STATEN ISLAND'. 

Each borough is further divided into community districts: 

> "The community boards of the New York City government are the appointed advisory groups of the community districts of the five boroughs. There are currently 59 community districts: twelve in Manhattan, twelve in the Bronx, eighteen in Brooklyn, fourteen in Queens, and three in Staten Island." Wikipedia, https://en.wikipedia.org/wiki/Community_boards_of_New_York_City   

Most of the boroughs also have additional Joint Interest Areas (JIA).  Which makeup the additional Community Board areas included in our Community Complaints database. You can ignore these for this section.  But if you're interested, you learn more about them at these Wikipedia pages:  
>  https://en.wikipedia.org/wiki/Community_boards_of_Queens,   
https://en.wikipedia.org/wiki/Community_boards_of_Brooklyn,  
https://en.wikipedia.org/wiki/Community_boards_of_Manhattan,   https://en.wikipedia.org/wiki/Community_boards_of_Staten_Island,  
  https://en.wikipedia.org/wiki/Community_boards_of_the_Bronx
  


In [18]:
comm_district_df = pd.read_csv('data/New_York_City_Population_By_Community_Districts.csv')

In [19]:
comm_district_df.head()

Borough  CD Number                             CD Name  1970 Population  \
0   Bronx          1    Melrose, Mott Haven, Port Morris           138557   
1   Bronx          2               Hunts Point, Longwood            99493   
2   Bronx          3       Morrisania, Crotona Park East           150636   
3   Bronx          4       Highbridge, Concourse Village           144207   
4   Bronx          5  University Hts., Fordham, Mt. Hope           121807   

   1980 Population  1990 Population  2000 Population  2010 Population  
0            78441            77214            82159            91497  
1            34399            39443            46824            52246  
2            53635            57162            68574            79762  
3           114312           119962           139563           146441  
4           107995           118435           128313           128200

## Analysing community district complaints per resident

Before we can run analysis on individual community district complaints we need to create a new `DataFrame` that combines the community board and district names with the 2010 district populations with our `complaints_df` data.  To merge these two data sets with a one-to-one join they need to share a common data column.  Both data sets include a community district id for each row, but they aren't using the same identifiers.  To do a one-to-one join you'll need to add a 'Community Board' field to `comm_district_df` which matches the format used in `complaints_df`.

`complaints_df`:

| Borough | Community Board |  
|------- |------- |  
| Bronx	 |	08 BRONX |

`comm_district_df`:   
  
| Borough | CD Number |	
|------- |------- |
| Bronx	 |	8 |	


### B4:  _Community board, district names  & 2010 populations_ (5 marks)
Create and return a `DataFrame` with columns for 'Community Board', 'CD Name', and '2010 Population'.  See table below which provides an example for the first row:

| &nbsp; | Community Board  |	CD Name | 2010 Population |
|------- |------- |------------| ------------| 
|0 | 01 BRONX |	Melrose, Mott Haven, Port Morris |	91497 |


*Hint: you might want to check out [pandas.Series.str.zfill](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.str.zfill.html)*

In [20]:
def b4():
    """Returns 'community board','cd name''2010 population' columns joined. One column(community board) added to comm_district_df"""
    df1 = pd.DataFrame(comm_district_df)
    df2 = pd.DataFrame(complaints_df)
    df1["Community Board"] = df2["Community Board"]
    df1["Community Board"]= df1["Community Board"].str.pad(15, side ='left')
    df3 = pd.concat([df1, df2], ignore_index=True, sort=False)
    
    df4 = df3[['Community Board','CD Name','2010 Population']]
    return df4

In [21]:
cd_pop = b4()
cd_pop.head()

Community Board                             CD Name  2010 Population
0         08 BRONX    Melrose, Mott Haven, Port Morris          91497.0
1         08 BRONX               Hunts Point, Longwood          52246.0
2         08 BRONX       Morrisania, Crotona Park East          79762.0
3      14 BROOKLYN       Highbridge, Concourse Village         146441.0
4         12 BRONX  University Hts., Fordham, Mt. Hope         128200.0

### B5: _Highest & lowest  noise complaints / 100 residents?_ (10 marks)

I'm looking for housing in New York City but I don't want to live in a noisy area.  

Create two `DataFrame`s one with the 10 community districts with the most noise related complaints per 100 residents and another with the 10 community districts with the least noise related complaints per 100 residents.  

Your function should return a `tuple` containing the two data frames.  The data frames will be displayed using the `class display(object)` function (as used in the *Python Data Science Handbook*).  See tables below which provide examples for the first rows.

df_most  
  
| &nbsp; | Community Board  |	CD Name | Complaints / 100 Residents |  
|------- |------- |------------|------------|  
| 34 |	05 MANHATTAN |	Midtown Business District |	5.59 |  


df_least 
  
| &nbsp; | Community Board  |	CD Name | Complaints / 100 Residents |  
|------- |------- |------------|------------|   
| 45 |	04 QUEENS |	Elmhurst, South Corona |	1.25 |  

In [138]:
def b5():
    """Returns 10 community districts with the most noise related complaints per 100 residents and another with the 10 community districts with the least noise related complaints per 100 residents"""
    
    df1 = pd.DataFrame(comm_district_df)
    df2 = pd.DataFrame(complaints_df)
    name = df1['CD Name']
    df3 = pd.merge(df2,name, left_index=True, right_index=True)
    noise = df3[df3['Complaint Type'].str.contains('[Nn]oise')]
    
    dup = pd.DataFrame(noise['Community Board'].value_counts().values, index=noise['Community Board'].value_counts().index, columns=['Complaints / 100 Residents'])
    
    return dup/100

In [139]:
# Display Most and Least Noise complaints below
b5()
#df_most, df_least = b5()
#display('df_most', 'df_least')

Complaints / 100 Residents
01 STATEN ISLAND                        0.04
12 BROOKLYN                             0.03
02 MANHATTAN                            0.03
07 QUEENS                               0.03
07 BROOKLYN                             0.02
12 BRONX                                0.02
10 BROOKLYN                             0.02
07 MANHATTAN                            0.02
03 QUEENS                               0.02
01 BRONX                                0.01
16 BROOKLYN                             0.01
11 MANHATTAN                            0.01
12 QUEENS                               0.01
06 BROOKLYN                             0.01
01 BROOKLYN                             0.01
01 QUEENS                               0.01
04 MANHATTAN                            0.01
14 BROOKLYN                             0.01
18 BROOKLYN                             0.01
05 BRONX                                0.01
06 QUEENS                               0.01
08 BRONX                                0.01
04 QUEENS                               0.01
01 MANHATTAN                            0.01
09 BRONX                                0.01
03 BRONX                                0.01
09 BROOKLYN                             0.01

### B6:  _Which community districts are the most popular ?_ (10 marks)

I'd also like to know which areas are the most popular. Create and return a `DataFrame` with the 10 community districts which have had the greatest % increase in population from 1970 to 2010.  See table below which provides an example for the first row.

| &nbsp; | Community Board  |	CD Name | 1970 Population | 2010 Population | Pop Increase %|  
|------- |------- |------------|------------| ------------|------------| 
| 30 | 01 MANHATTAN | Battery Park City, Tribeca | 7706 | 60978 | 87.36 |  

In [20]:
def b6():
    """Returns 10 community districts which have had the greatest % increase in population from 1970 to 2010 """
    df1 = pd.DataFrame(comm_district_df)
    diff = df1.apply(lambda row: (row.iloc[7]-row.iloc[3])/row.iloc[7]*100, axis=1)
    df1['Pop Increase %'] = diff.sort_values(ascending=False)
    df1['Community Board'] = df1.sort_values(by = ['Community Board'], ascending=True)
    col2 = df1['CD Name']
    col3 = df1.sort_values(by = ['1970 Population'], ascending=True)
    col4 = df1.sort_values(by = ['2010 Population'], ascending=True)
    
    return col3[['Community Board','CD Name','1970 Population','2010 Population','Pop Increase %']][:10]
    

In [21]:
b6()

Community Board                             CD Name  1970 Population  \
30       Manhattan          Battery Park City, Tribeca             7706   
34       Manhattan           Midtown Business District            31076   
58   Staten Island   Tottenville, Woodrow, Great Kills            72815   
33       Manhattan                    Chelsea, Clinton            83601   
31       Manhattan             Greenwich Village, Soho            84337   
9            Bronx  Throgs Nk., Co-op City, Pelham Bay            84948   
57   Staten Island        New Springville, South Beach            85985   
43          Queens                 Sunnyside, Woodside            95073   
24        Brooklyn        Coney Island, Brighton Beach            97750   
55          Queens        The Rockaways, Broad Channel            98228   

    2010 Population  Pop Increase %  
30            60978       87.362655  
34            51673       39.860275  
58           160209       54.549994  
33           103245       19.026587  
31            90016        6.308878  
9            120392       29.440494  
57           132003       34.861329  
43           113200       16.013251  
24           104278        6.260189  
55           114978       14.568004